packages/react-components/react-dialog/src/components/Dialog/useDialog.ts

```diff
@@ -9,20 +15,181 @@ import type { DialogProps, DialogState } from './Dialog.types';
  * before being passed to renderDialog_unstable.
  *
  * @param props - props from this instance of Dialog
- * @param ref - reference to root HTMLElement of Dialog
  */
-export const useDialog_unstable = (props: DialogProps, ref: React.Ref<HTMLElement>): DialogState => {
+export const useDialog_unstable = (props: DialogProps): DialogState => {
+  const { children, overlay, modalType = 'modal', onOpenChange } = props;
+
+  const [trigger, content] = childrenToTriggerAndContent(children);
+
+  const [open, setOpen] = useControllableState({
+    state: props.open,
+    defaultState: props.defaultOpen,
+    initialState: false,
+  });
+
+  const overlayShorthand = resolveShorthand(overlay, {
+    required: modalType !== 'non-modal',
+    defaultProps: {
+      'aria-hidden': 'true',
+    },
+  });
+
+  const requestOpenChange = useEventCallback((data: DialogRequestOpenChangeData) => {
```

Comment:
nit: not a big fun of using "request" prefix as there are clear actions for be executed... Something like `toggleOpen`/`setOpen`/`dispatchOpen` sounds better. But it's all IMO